Fix all py3 api tests

mvdbeek opened the pull request:
~~We're almost there, locally only the history import/export fails, and the test_workflow_stability test.~~
I'm running tests using `GALAXY_TEST_TOOL_CONF=config/tool_conf.xml.sample,test/functional/tools/samples_tool_conf.xml pytest -v test/api`.
Also fixes some remaining issues with webob/paste and mako. This feels almost functional!

xref: #1715

mvdbeek committed:
Replace paste.httpexception with webob.exc classes

mvdbeek committed:
Fix make_file method of patched cgi_FieldStorage

This copies the behavior from webob.compat.cgi_FieldStorage,
except we use a NamedTemporaryFile instead of TemporaryFile.

mvdbeek committed:
Pass bytestring to hash function

mvdbeek committed:
py3: Fix history import/export tool

 - Open history tarfile in binary mode
 - Fix base64 encoding in history import tool
 - test history export more extensively

mvdbeek committed:
Fix binary file check

mvdbeek committed:
Fix cgi_FieldStorage monkey patch

self.__file is defined in cgi.FieldStorage, so that is the class
that needs to be patched. Followup to a65cd29c366035fb8b26e2274046e8c96e3cab33

mvdbeek committed:
py3: Fix dataset collection test

mvdbeek committed:
py3: fix check_html

mvdbeek committed:
py3: Fix authenticate api test

mvdbeek committed:
Cast to bytestring for is_binary check

mvdbeek committed:
py3: Fix raw dataset display

mvdbeek committed:
py3: Fix catching malformed IDs

mvdbeek committed:
py3: Fix tool_data API

mvdbeek committed:
py3: Fix upload jobs API tests

mvdbeek committed:
py3: Fix library tests

mvdbeek committed:
py3: Send bytestring in various datatatypes' display_data methods

mvdbeek committed:
py3: Cast map objects to lists when serializing tools to json

mvdbeek committed:
py3: Fix getting first webhook

mvdbeek committed:
py3: Don't rely on order of outputs, instead match by output name

mvdbeek committed:
py3: Use reponse.text instead of reponse.content for decoded text

mvdbeek committed:
Make random_lines reproducible between py2 and py3

random.seed has changed in python 3, and while there's backwards
compatibility using `version=1`, random.randint and random.sample have
also changed. The most striaghtforward change without adding additional
dependencies then was to copy those methods from the python 2.7 source
code.

mvdbeek committed:
py3: Fix the remaining mako issues

mvdbeek committed:
Fix redirection unit test

nsoranzo commented on lib/galaxy/tools/imp_exp/unpack_tar_gz_archive.py:
```diff
@@ -64,7 +64,7 @@ def unpack_archive(archive_file, dest_dir):
     Unpack a tar and/or gzipped archive into a destination directory.
     """
-    archive_fp = tarfile.open(archive_file, mode='r')
+    archive_fp = tarfile.open(archive_file, mode='rb')
```
`rb` is not listed among the allowed modes in https://docs.python.org/3/library/tarfile.html#tarfile.open , is that necessary?

mvdbeek replied:
Thanks, it is not, there were and are still other issues with the import/export , some of which I fixed in https://github.com/galaxyproject/galaxy/pull/6561/commits/025167546202591524a41ae1f8ceddb859cbac37

mvdbeek committed:
fix is_binary

mvdbeek committed:
py3: Fix redirection

mvdbeek committed:
Fix webob.compat make_file patching for py2

jmchilton commented on tools/filters/random_lines_two_pass.py:
```diff
@@ -7,16 +7,69 @@
+            if isinstance(population, list):
+                raise
+            return sample(tuple(population), k)
+    return result
```
Gross... but needed - thanks for figuring that out!

mvdbeek committed:
py3: Fix DataParameter test data upload

mvdbeek committed:
py3: fix safe_eval

mvdbeek committed:
py3: Fix mulled resolver

mvdbeek committed:
py3: fix sha1 hashing

mvdbeek committed:
py3: Fix exception handling in wait_on method

mvdbeek committed:
py3: fix test_scripts integration test

mvdbeek committed:
py3: Fix upload_configuration_options integration test

mvdbeek committed:
py3: Fix test_fetch_zip_to_folder test

mvdbeek committed:
Explicitly sort workflow step edges

This seems to be necessary so that the workflow_stability test passes
under python 2 and python3, but does change the result under python 2,
hence the updated workflows (produced by uploading and downloading the
workflow to an instance running with this commit).

mvdbeek commented on lib/galaxy/workflow/steps.py:
```diff
@@ -35,7 +35,7 @@ def order_workflow_steps(steps):
         steps.sort(key=lambda _: math.sqrt(_.position['left'] ** 2 + _.position['top'] ** 2))
     try:
-        edges = edgelist_for_workflow_steps(steps)
+        edges = sorted(edgelist_for_workflow_steps(steps))
```
@jmchilton does this make sense to you ? I'm a little concerned that by updating the test workflows I have bent the test to pass for the sake of passing ...

jmchilton replied:
Hmm... such a good question. The fact that sorted is needed means the previous test was somehow depending on the particular hashing or random lib behavior of Python 2.7 I assume - which is less than ideal. I'd like to establish a "correct order" for the steps in a workflow so ``order_index`` is preserved for structurally similar workflows across saves - the test still passes so it is preserved - its just different than it was previously. ``topsort`` using ``random`` so I guess I shouldn't be surprised I'm failing to establish a correct order - the test probably stinks.

mvdbeek replied:
I don't think `topsort` itself is the issue. I tried this in isolation using the `pairlist` I intercepted form python 2, it gives consistent results between py2 and py3. I think something is different in `edgelist_for_workflow_steps`, perhaps the order in which `step.input_connections` are listed ?

mvdbeek commented:
This is ready for review (while I set up a jenkins job for python 3).

mvdbeek commented on lib/galaxy/tools/imp_exp/imp_history_from_archive.xml:
```diff
@@ -1,7 +1,11 @@
-  <command>python '$__tool_directory__/unpack_tar_gz_archive.py' '${ str( $__ARCHIVE_SOURCE__ ).encode( 'base64' ) }' '${ str( $__DEST_DIR__ ).encode( 'base64' ) }' --$__ARCHIVE_TYPE__ --encoded</command>
+  <command>#from base64 import b64encode#
+python '$__tool_directory__/unpack_tar_gz_archive.py'
+'${ b64encode(bytearray(str($__ARCHIVE_SOURCE__), encoding="utf-8")).decode('utf-8')}'
```
The need to wrestle with b64encode, bytearray and string decoding unfortunately made we realise that of course @jmchilton was right at the GCC , we probably need to keep supporting python 2 for a very long time if we want to remain reproducible. I think it's a small minority of tools though that will fail if not updated:(.

jmchilton commented on lib/galaxy/jobs/runners/state_handlers/_safe_eval.py:
```diff
@@ -19,7 +19,9 @@
-def _check_name(ast_node, allowed_variables=[]):
+def _check_name(ast_node, allowed_variables=None):
+    if allowed_variables is None:
+        allowed_variables = []
```
I need to stop doing that - sorry. Thanks for the fix!

jmchilton commented on lib/galaxy/webapps/galaxy/api/tools.py:
```diff
@@ -174,6 +174,8 @@ def test_data(self, trans, id, **kwd):
             if isinstance(obj, odict):
                 return dict(obj)
+            elif isinstance(obj, map):
+                return list(obj)
```
The Java programmer in me was confused by this - but I guess ``map`` is the iterable result type of mapping over a list so this makes sense.

jmchilton commented:
Thanks a million @mvdbeek - amazing!

jmchilton commented:
Saw this in the Selenium test logs for dev, related?

```
Traceback (most recent call last):
  File "/galaxy_venv/local/lib/python2.7/site-packages/paste/httpserver.py", line 1085, in process_request_in_thread
    self.finish_request(request, client_address)
  File "/usr/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/lib/python2.7/SocketServer.py", line 652, in __init__
    self.handle()
  File "/galaxy_venv/local/lib/python2.7/site-packages/paste/httpserver.py", line 459, in handle
    BaseHTTPRequestHandler.handle(self)
  File "/usr/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/galaxy_venv/local/lib/python2.7/site-packages/paste/httpserver.py", line 454, in handle_one_request
    self.wsgi_execute()
  File "/galaxy_venv/local/lib/python2.7/site-packages/paste/httpserver.py", line 304, in wsgi_execute
    self.wsgi_start_response)
  File "/galaxy_venv/local/lib/python2.7/site-packages/paste/urlmap.py", line 216, in __call__
    return app(environ, start_response)
  File "/galaxy/lib/galaxy/web/framework/middleware/static.py", line 48, in __call__
    app.cache_control(max_age=int(self.cache_seconds))
AttributeError: 'FileApp' object has no attribute 'cache_control'
```

mvdbeek commented:
Yes, I'll have a look!
